pkg/lease/lease_test.go

```diff
@@ -0,0 +1,285 @@
+	srv := grpc.NewServer()
+	listener := bufconn.Listen(1e6)
+
+	metricServer := &testMetricService{
+		latency:        leaseOpts.RetryPeriod / 10,
+		frequencyLimit: leaseOpts.UpdatePeriod,
+	}
+	monitoring_pb.RegisterMetricServiceServer(srv, metricServer)
+
+	go func() { srv.Serve(listener) }()
+	defer srv.Stop()
+
+	// Run multiple leases simultaniously and wait for a condition violation for the given time.
+	// If there is a leader most of the time and never more than one leader, everything is fine.
+	ctx, cancel := context.WithTimeout(context.Background(), duration)
+	defer cancel()
+
+	bufDialer := func(context.Context, string) (net.Conn, error) {
+		return listener.Dial()
+	}
+	client, err := monitoring.NewMetricClient(ctx,
+		option.WithoutAuthentication(),
+		option.WithGRPCDialOption(grpc.WithInsecure()),
+		option.WithGRPCDialOption(grpc.WithContextDialer(bufDialer)),
+	)
+	if err != nil {
+		t.Fatalf("creating metric client failed: %s", err)
+	}
+
+	var (
+		resource = &monitoredres_pb.MonitoredResource{
+			Type:   "prometheus_target",
+			Labels: map[string]string{"instance": "replica-key"},
+		}
+		intervals = newReplicaIntervals(leaseOpts, time.Now())
+		// Control number of concurrent replicas via channel.
+		replicaCh = make(chan struct{}, 100)
+	)
+	hook := func(name string, start, end time.Time, valid bool) {
+		if valid {
+			intervals.add(name, start, end)
+		}
+	}
+
+Loop:
+	for i := 0; ; i++ {
+		select {
+		case <-ctx.Done():
+			break Loop
+		case replicaCh <- struct{}{}:
+		}
+		// Keep producing new replica IDs. In practice IDs would be recycled but
+		// this makes it easier to debug history.
+		name := fmt.Sprintf("replica-%d", i)
+
+		logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stdout))
+		logger = log.With(logger, "ts", log.DefaultTimestampUTC, "replica", name)
+
+		lease, err := New(nil, "test-project", client, resource, leaseOpts)
+		if err != nil {
+			t.Fatalf("Failed to create lease: %s", err)
+		}
+		lease.Register(func(start, end time.Time, ok bool) {
+			hook(name, start, end, ok)
+		})
+		// Make replicas die after a random amount of time unless a fixed lifetime is given.
+		if replicaLifetime == 0 {
+			replicaLifetime = time.Duration(rand.Intn(int(50 * leaseOpts.ExtensionPeriod)))
+		}
+		t.Logf("starting %s with lifetime %s", name, replicaLifetime)
+
+		ctx, cancel := context.WithTimeout(ctx, replicaLifetime)
+		go func() {
+			lease.Run(ctx)
+			cancel()
+			// Free replica spot on exit.
+			<-replicaCh
+		}()
+	}
+
+	t.Log(intervals.intervals())
+
+	if err := intervals.validate(); err != nil {
+		t.Fatal(err)
+	}
+	return intervals.intervals()
+}
+
+func TestLease(t *testing.T) {
+	opts := &Options{
+		UpdatePeriod:    10 * time.Millisecond,
+		RetryPeriod:     1 * time.Millisecond,
+		ExtensionPeriod: 30 * time.Millisecond,
+	}
+
+	t.Run("one-lease-interval", func(t *testing.T) {
+		// This test ensures that the lease is not flip-flopping between participants if they
+		// remain alive.
```

Comment:
This test is asserting more than that right? The `if len(intervals) != 1` looks to be checking that only one replica owned the lease during the duration of the test.